processor/src/fast/mod.rs

```diff
@@ -579,25 +751,62 @@ impl FastProcessor {
         call_node: &CallNode,
         current_node_id: MastNodeId,
         program: &Program,
-        current_forest: &MastForest,
+        current_forest: &Arc<MastForest>,
         continuation_stack: &mut ContinuationStack,
         host: &mut impl AsyncHost,
     ) -> Result<(), ExecutionError> {
+        self.check_extract_trace_state(
+            NodeExecutionPhase::Start(current_node_id),
+            continuation_stack,
+            current_forest,
+        );
+
         // Execute decorators that should be executed before entering the node
         self.execute_before_enter_decorators(current_node_id, current_forest, host)?;
 
         let err_ctx = err_ctx!(current_forest, call_node, host);
 
-        // Corresponds to the row inserted for the CALL or SYSCALL
-        // operation added to the trace.
-        self.clk += 1_u32;
-
         // call or syscall are not allowed inside a syscall
         if self.in_syscall {
             let instruction = if call_node.is_syscall() { "syscall" } else { "call" };
             return Err(ExecutionError::CallInSyscall(instruction));
         }
 
+        // In tracing mode, record the control block on the block stack
+        if self.trace_state_builder.is_some() {
```

Comment:
maybe an if-let?

Reply:
We need to do this for borrow-checker reasons - basically an `if let Some(trace_state_builder) = &mut self.trace_state_builder` would mutably borrow all of `self`, which then prevents us from calling e.g. `self.stack_depth()` (which attempts to immutably borrow it). But this will be naturally cleaned up soon when we implement #2089.